hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/states/Node2PipelineMap.java

```diff
@@ -71,6 +71,10 @@ public synchronized void addPipeline(Pipeline pipeline) {
       UUID dnId = details.getUuid();
       dn2ObjectMap.computeIfAbsent(dnId, k -> ConcurrentHashMap.newKeySet())
           .add(pipeline.getId());
+      dn2ObjectMap.computeIfPresent(dnId, (k, v) -> {
+        v.add(pipeline.getId());
```

Comment:
why do we need to add the dn->pipelineId in 75 since line 73 already add it?

Reply:
Line 73 is adding v to new keyset when key is absent and line 75 is adding v to the existing keyset. Reason why we don't have to have extra if here is computeIfAbsent and computeIfPresent are both judging if key is present or not. So we don't have to see if Map.get(k) is null.

Reply:
Correct me if I'm wrong, dn2ObjectMap.computeIfAbsent(dnId, k -> ConcurrentHashMap.newKeySet()) ensures for a dnId we will have a KeySet().

Then the returned value (KeySet) allows line 73 to continue the operation to add pipeline id with a .add(pipeline.getId).

JDK document can be found here: https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#computeIfAbsent-K-java.util.function.Function-

Reply:
The document says existing value(keyset) will be returned. So the line 74-75 is not necessary.
 
Returns:
the current (existing or computed) value associated with the specified key,

A good example can be found here: https://stackoverflow.com/a/39268896

Reply:
I see. Thanks for the tipping. Based on the example here, Shall I update the removePipeline method? I'm making a change and please take a look.